File: BEmu_cpp/src/MarketDataRequest/MarketElementSubscriptionCancelReason.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/MarketDataRequest/MarketElementSubscriptionCancelReason.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "MarketDataRequest/MarketElementSubscriptionCancelReason.h"
#include "MarketDataRequest/MarketElementString.h"
#include "MarketDataRequest/MarketElementInt.h"

#include "Types/IndentType.h"

namespace BEmu
{
	namespace MarketDataRequest
	{
		MarketElementSubscriptionCancelReason::MarketElementSubscriptionCancelReason() :
            _source(new MarketElementString("source", "Session")),
            _errorCode(new MarketElementInt("errorCode", 0)),
            _category(new MarketElementString("category", "CANCELED")),
            _description(new MarketElementString("description", "Subscription canceled")),
            _subCategory(new MarketElementString("subcategory", ""))
		{
		}

		MarketElementSubscriptionCancelReason::~MarketElementSubscriptionCancelReason()
		{
		}

		Name MarketElementSubscriptionCancelReason::name() const
		{
			return Name("reason");
		}

		size_t MarketElementSubscriptionCancelReason::numValues() const
		{
			return 1;
		}

		size_t MarketElementSubscriptionCancelReason::numElements() const
		{
			return 5;
		}

		bool MarketElementSubscriptionCancelReason::isNull() const
		{
			return false;
		}

		bool MarketElementSubscriptionCancelReason::isArray() const
		{
			return false;
		}

		bool MarketElementSubscriptionCancelReason::isComplexType() const
		{
			return true;
		}

		std::ostream& MarketElementSubscriptionCancelReason::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << "reason = {" << std::endl;
			this->_source->print(stream, level + 1, spacesPerLevel);
			this->_errorCode->print(stream, level + 1, spacesPerLevel);
			this->_category->print(stream, level + 1, spacesPerLevel);
			this->_description->print(stream, level + 1, spacesPerLevel);
			this->_subCategory->print(stream, level + 1, spacesPerLevel);

			stream << tabs << '}' << std::endl;

			return stream;
		}
	}
}

File: BEmu/MarketDataRequest/MarketElementSubscriptionCancelReason.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="MarketDataRequest/MarketElementSubscriptionCancelReason.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class MarketElementSubscriptionCancelReason : Element
    {
        private readonly MarketElementString _source, _category, _description, _subCategory;
        private readonly MarketElementInt _errorCode;

        internal MarketElementSubscriptionCancelReason()
        {
            this._source = new MarketElementString("source", "Session");
            this._errorCode = new MarketElementInt("errorCode", 0);
            this._category = new MarketElementString("category", "CANCELED");
            this._description = new MarketElementString("description", "Subscription canceled");
            this._subCategory = new MarketElementString("subcategory", "");
        }

        public override Name Name { get { return new Name("reason"); } }
        public override int NumElements { get { return 5; } }
        public override int NumValues { get { return 1; } }
        public override bool IsArray { get { return false; } }
        public override bool IsComplexType { get { return true; } }
        public override bool IsNull { get { return false; } }

        public override object this[int index] { get { return null; } }

        public override Element this[string name]
        {
            get
            {
                foreach (var item in this.Elements)
                {
                    if (item.Name.ToString() == name)
                        return item;
                }
                return base[name];
            }
        }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._source;
                yield return this._errorCode;
                yield return this._category;
                yield return this._description;
                yield return this._subCategory;
            }
        }

        internal override StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}reason = {{{1}", tabs, Environment.NewLine);
            foreach (var item in this.Elements)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);
            return result;
        }
    }
}
